Bring context logger with edge request ID

lzap opened the pull request:
Logging without correlation information is not the best practice for services under heavy load since there is nothing you can do to find logs which belong to each other. The standard practice is to add some sort of correlation id or request id. On our platform, that is the X-Rh-Edge-Request-Id which is set by 3scale and can be propagated through services. Since our service (launch) does exactly that, I would love to see correlated logs not only within our app, but also across services we use.

Image builder currently use logrus global logger directly which does not allow for context propagation. The standard practice is to put logger into the context and use it instead for all application logging. Instead:

    logrus.Info("msg")

you should use (in case of Echo context):

    ctx.Logger().Info("msg")

Luckily the two have the same interface so this should be search and find replace for the whole codebase. In my patch, I only change one line for the purposes of easy review. Unfortunately, for other types (or libraries) the context must be passed into all functions - this is a bit of work that can be gradually done. Also, echo context can be converted to Go standard library context if needed to be passed into other libraries. In our app, we have context logging for all important libraries (SQL, AWS SDK).

The way I approached is to use the same what you guys do in osbuild-composer - I am bringing a small type EchoLogger that works as a proxy between Echo logger and Logrus. Then I am telling to Echo to use that loger, also to hide the startup banner (would that break in ELK fonts?) and I am adding Echo logging middleware that allows a nice way of capturing all important structured fields. Finally, when TRACE or DEBUG logging level is configured, I put Echo in debug mode which has some advantages (e.g. output is JSON human formatted like below). This is how a log looks like:

```
time="2023-02-16T12:06:35+01:00" level=info msg="Processed request GET /404" func=main.main.func1 file="/Volumes/SSD/Work/image-builder/cmd/image-builder/main.go:116" latency_ms=0 method=GET request_id=123456789 status=404 uri=/404
```

As you can see, this is a behavior change - every request now have an INFO message with some details about the request, latency, status and error message (if any). This allows nice and quick searching in ELK for example - show me just errors etc.

The bread and butter of this patch is extracting the edge request ID from headers and putting it as field named "request_id". When not present (e.g. in ephemeral) a new random string is generated so correlation is still possible.

```
# curl -H "X-Rh-Edge-Request-Id: 123456789" http://127.0.0.1:8086/404
{
  "errors": [
    {
      "detail": "Not Found",
      "title": "404"
    }
  ]
}
```

I needed to add echo/middleware package, it brings few dependencies. I hope that is not the issue, I really hope you will like this change which brings better logging experience to all of us. Cheers!

lzap commented on cmd/image-builder/main.go:
```diff
@@ -87,6 +90,37 @@
+			if values.Error != nil {
+				fields["error"] = values.Error
+			}
+			logrus.WithFields(fields).Infof("Processed request %s %s", values.Method, values.URI)
```
False alert, but I can add `strings.Replace` to make the bot happy.

croissanne replied:
ye let's make the bot happy :)

teg commented:
This makes a lot of sense in principle and sounds like it would make all of our lives a lot easier, though I'm not well-versed enough in the details to judge.

cc @diaasami

lzap commented:
> Introducing a request_id can make it harder since there are two IDs now, request_id and jobID, However, request_id can help when debugging a case with no jobID if this happens.

What you mean harder? There is no limitation in the ELK stack on how many fields you can have. In fact in our app, we have multiple IDs logged alongside each log record:

* request_id (the same thing as in this patch)
* trace_id (OpenTelemetry tracing id)
* account_id (our own internal account table id)
* account_number/org_id (business account IDs)
* job_id (uuid of the job from our job queue - in case this is a background worker)

Different pods may have different fields, in our case API pods do have request_id and trace_id but worker jobs (job task queue executors) do have job_id only (there is no request). All of them do have account information, that can be easily added too as a middleware in a similar way.

If you want to test how it feels to work with our logs, just search for these fields in ELK stage, just make sure to start your query with `@log_group: "provisioning-cloudwatch-stage" and @log_stream: "pbapi"` if you want to see logs from our app only. Here is a little screenshot:

![image](https://user-images.githubusercontent.com/49752/219415786-ba713f3b-093c-49ff-8edd-bb6e4a006e1b.png)

croissanne commented on HACKING.md:
```diff
@@ -0,0 +1,21 @@
+# Image Builder developer notes
+
+Random notes that can boost your onboarding experience.
+
+## Running the project without composer
+
+It is possible to provide fake composer connection in order to start the service:
+
+    PGHOST=nuc
+    PGDATABASE=database
+    PGUSER=user
+    PGPASS=password
+    COMPOSER_CLIENT_ID=1
+    COMPOSER_TOKEN_URL=http://localhost
+    COMPOSER_OFFLINE_TOKEN=1
+    DISTRIBUTIONS_DIR=distributions
+    LOG_LEVEL=trace
+
+Then build and run the project, or just:
+
+    make run
```
Can we unify this and test/README.md?

croissanne commented on cmd/image-builder/main.go:
```diff
@@ -87,6 +90,37 @@ func main() {
 	echoServer := echo.New()
+	echoServer.HideBanner = true
+	echoServer.Logger = common.Logger()
+	echoServer.Use(middleware.RequestLoggerWithConfig(middleware.RequestLoggerConfig{
```
So all other webserver config is currently located in v1/server.go in Attach() (things like errorhandler and Binder, stuff that's common between all echo groups).

I think it makes sense to extract all server-wide config (like logging, binder, errorhandler), since for v2 we'll need that in a common place. But let's do that in a followup.

croissanne reviewed (changes requested):
Let's make the bot happy, otherwise lgtm.

lzap commented:
Fixed all of that plus I also replaced `logrus` with `ctx.Logger()` in hopefully all places where I could do this without modifying any function parameters. There are few other places where we need to pass `ctx` but I do not want to do this now, in this PR. These are typically error messages which will get to the user and into logs eventually through the generic "catch" error code.

lzap commented:
Updated commit messages.

For the record, I thought for a moment I would also contribute org_id and account_number fields, but it looks like you have some public API plans and identity gets parsed down the stack so it is not as easy as:

```go
			id, ok := identity.Get(c.Request().Context())
			if ok {
				fields["org_id"] = id.Identity.OrgID
				fields["account_number"] = id.Identity.AccountNumber
			}
```

Leaving this on the table for now.

croissanne commented:
Took the liberty to rebase it
